### DISCUSSION TITLE
How do I override component styles?

### DISCUSSION THREAD

**Discussion by AaronBeaudoin** [tailwindlabs/tailwindcss#6375] [Help]
I've created a basic `BaseButton.vue` button component:

```vue
<template>
  <button class="flex justify-center items-center h-10 px-8 rounded-full
  bg-black hover:bg-gray-800 text-white font-black uppercase tracking-wide">
    <slot></slot>
  </button>
</template>
```

Easy enough. Now I want to override it in another `LargeButton.vue` component (or maybe just wherever I use it in my application) to tweak the appearance of the button:

```vue
<template>
  <BaseButton class="h-12 px-12 text-lg"><slot></slot></BaseButton>
</template>
```

Yikes. Now I have a problem. My HTML will end up with both a `h-10` and `h-12` class, and both a `px-8` and `px-12` class; and I haven't a clue which one will actually end up appearing first in the actual CSS.

A bit of searching seems to indicate the recommended solution is to extract all of my base button styles using `@apply`. Here's my new `BaseButton.vue` component:

```vue
<template>
  <button class="button">
    <slot></slot>
  </button>
</template>

<style lang="postcss">
@layer components {
  .button {
    @apply flex justify-center items-center rounded-full font-black uppercase tracking-wide;
    @apply h-10 px-8 bg-black hover:bg-gray-800 text-white;
  }
}
</style>
```

Well that didn't work. I get the error:

```
`@layer components` is used but no matching `@tailwind components` directive is present.
```

More searching. It seems like this is basically telling me to move my `.button` class into my `index.css` file:

```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .button {
    @apply flex justify-center items-center rounded-full font-black uppercase tracking-wide;
    @apply h-10 px-8 bg-black hover:bg-gray-800 text-white;
  }
}
```

It works, but this sucks. Now my button styles live in a totally different part of my project than my button component. My `BaseButton.vue` component now looks like this:

```vue
<template>
  <button class="button"><slot></slot></button>
</template>
```

At first glance, anyone would look at this and be like "What does the `.button` class actually do, and where is it even defined?" This component looks essentially useless at this point.

So what's the right way to do this? Please don't tell me the solution is to just write `<button class="button">` everywhere and stop using a component, because that doesn't address the fact that my styles are now somewhere else in my code, and it really feels like an entire step backwards.

Also, can we talk about the fact that the docs seem to suggest that `@apply` is for when you want to avoid style duplication without using a component? In this case however, it's clearly necessary even when my button styles are extracted into a component, because otherwise I can't reliably override any of the styles anywhere I use the component.

I don't want to sound like I'm bashing Tailwind. I absolutely love it. In this particular case, however, it's not clear to me what the philosophy is. Are components supposed to be treated as totally closed so their CSS can't be overridden?? Is my component CSS just supposed to be relocated to my CSS files in order to make them overridable? Help!

**Comment by philwolstenholme** [comment]
Hi :)

> Are components supposed to be treated as totally closed so their CSS can't be overridden

That's the way I treat things. In the past, with older CSS methodologies we'd end up in a mess because we'd have layers and layers of CSS added over time to override things, and no one would fully understand the codebase once it was more than a year or two old. We'd end up with 'append-only CSS' (https://css-tricks.com/oh-no-stylesheet-grows-grows-grows-append-stylesheet-problem)

I can't speak for every Tailwind user, but avoiding overrides has changed my work life. I've moved towards an approach where every styling difference is handled by props and arguments to components, not the cascade. This makes it much less mentally taxing to work out why something looks the way it does.

Here's an example of a Vue button component that uses dynamic classes to apply different padding and text size options to a button:

* https://codesandbox.io/s/dazzling-hopper-ph3uj?file=/src/components/HelloWorld.vue
* https://codesandbox.io/s/dazzling-hopper-ph3uj?file=/src/components/TwButton.vue

```vue
<ul class="space-y-4">
  <li><TwButton>I'm a regular button</TwButton></li>
  <li><TwButton size="lg">I'm a big button</TwButton></li>
  <li><TwButton size="sm">I'm a small button</TwButton></li>
  <li><TwButton padding="tight">I'm a tightly padded button</TwButton></li>
  <li><TwButton padding="loose">I'm a loosely padded button</TwButton></li>
</ul>
```

```vue
<template>
  <button
    class="bg-black text-white rounded font-bold shadow-lg select-none transition hover:scale-110 focus:scale-110"
    onClick="alert('Hello')"
    :class="{
      'text-xxl': this.size == 'lg',
      'text-xs': this.size == 'sm',
      'px-4 py-2': this.padding == 'md',
      'px-2 py-1': this.padding == 'tight',
      'px-8 py-4': this.padding == 'loose',
    }"
  >
    <slot>Button</slot>
  </button>
</template>
```

## What are the advantages of this approach?

- You can customise your component's appearance however you like
- You only ever need to check the component's source to see how it will ever look, you don't need to worry about the styles in `button.scss` being overridden by some random line in `shopping-cart.scss` or `header.scss`
- You enforce consistency across the site. No one can sneakily make the button use `font-size: 15px` in one place and `16px` everywhere else.
- It's easy to connect your components to things like Storybook to see all the possible visual options in one place
- Your teammates don't need to mentally hold the whole CSS codebase in their heads and mentally apply the cascade and specificity rules anymore as we have much fewer overriding selectors. You can still use this for styling things like rich-text/WYSIWYG text, but not for components

## What are the downsides?

- Tailwind means you no longer need to come up with BEM style names, but now you need to name a bunch of props!
- Renaming or removing an option means updating a bunch of files, but IDEs make this pretty painless 
- Probably some other things, but honestly, nothing that I can remember regularly getting in my way, I personally love this approach :)

**Reply by philwolstenholme** [reply]
There's also some discussion about ways to merge classes here, but I think the props approach is easiest and probably more appropriate for someone who's not a Tailwind power user yet! https://github.com/tailwindlabs/tailwindcss/discussions/1446

**Reply by AaronBeaudoin** [reply]
I have to say I'm not super excited about this approach.

- Components in any framework—including Vue—are designed to allow you to extend/compose/wrap other components, so they're not closed at all by nature. Treating styling as totally closed and "non-extendable" feels like a massive shortcoming and I don't feel like it aligns with how most developers build web applications. It really feels like a step backwards. (I recognize that this is largely due to the nature of CSS, not necessarily Tailwind, but I maintain that we should be looking to better align how we compose styles with how we compose the components they are associated with.)
- I don't know what IDE you're using, but in VS Code, my experience refactoring code in Vue component templates has been anything but "painless". I agree with @mariusmarais in [this](https://github.com/tailwindlabs/tailwindcss/discussions/1446#discussioncomment-620615) thread in the discussion you linked about merging classes. If you have a constantly expanding project that may need to go occasional rebranding, then the "props" approach is pretty much a nightmare.
- I agree with the need to enforce consistency across a site, but in basically every real world project I've ever worked on there have been weird edge cases where a component needed some tweaked styles. With this approach, you'd either have to duplicate the styles into a new component with some slight modifications, or bake a way to make that tweak into the existing component, in which case you've added logic for some specific use case that might not even make sense in the context of what the component itself does. This seems like a fatal flaw to me.

Overall, I feel like this is the "dirty secret" of Tailwind. The moment you need to start composing components and overriding existing styles (which happens in basically every project ever), the system kind of falls apart. I'm conflicted, because there's so much I love about Tailwind, but this particular issue is pretty painful.

**Reply by philwolstenholme** [reply]
I suppose this is probably a Vue/React/whatever-other-template-language problem rather than a utility-first CSS problem.

A central point of utility-first is that you compose styles with atomic single-purpose classes that are immutable, so we can't override them using the CSS cascade, or else the predictability that looking at a string of Tailwind classes and knowing exactly how that component will look is gone. How those strings of classes end up in the HTML that is sent to browsers is left up to the author. 

---

Looking back at your original post I think using `@apply` might work for you if you want to avoid the props-heavy approach. It looks like something might be wrong with your setup though as the issues you described are fixable :)

This message:

> `@layer components` is used but no matching `@tailwind components` directive is present.

Doesn't mean you have to do this:

```
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .button {
    @apply flex justify-center items-center rounded-full font-black uppercase tracking-wide;
    @apply h-10 px-8 bg-black hover:bg-gray-800 text-white;
  }
}
```

But it does mean that your Vue style block from `BaseButton.vue` isn't being processed in the same Postcss run/context as your `index.css` file is being processed by Postcss. That's why Tailwind can't find the `@tailwind components` directive. You should be able to fix this though. It'll be something to do with your Webpack setup I think, or perhaps to do with how `index.css` is being included in your app? Are you including it via Webpack with `import './index.css'` rather than using a `<style>` element in your HTML markup?

Quick edit: take my advice with a pinch of salt as it's been a while (a few years) since I've used Tailwind and Vue `<style>` blocks as I normally gravitate towards the props-heavy approach.

I found https://markus.oberlehner.net/blog/reusable-functional-vue-components-with-tailwind-css when I was looking for links to help you, but I don't think it describes what you're after - probably the opposite in fact! But it might be worth a read just to get a second opinion.

**Reply by AaronBeaudoin** [reply]
Actually, I think the specific issue I described is **not** fixable. See [this](https://tailwindcss.com/docs/adding-custom-styles#layers-and-per-component-css) part of the Tailwind docs.

**TLDR:** You can't use `@layer` in component styles, and unfortunately as a result:

> You lose the ability to control the precedence of your styles, but unfortunately that’s totally out of our control because of how these tools work.

This specifically is the issue I was trying to overcome, but it looks like that is officially not an option.

In response to our related discussion, I did a ton of research and initially tried the whole "class merging" approach, and while it was really cool (see the [`tailwind-merge`](https://github.com/dcastil/tailwind-merge) utility), it didn't play well with how Vue automatically merges the `class` attribute when it is reused outside of a component (see [this](https://github.com/dcastil/tailwind-merge/discussions/76) discussion). It also adds complexity to tooling and may noticeably affect performance in a large project (I haven't tested that though).

As a result, I gave the prop approach a try. While I still don't really like it, I can see some of the advantages. I'm starting to reach the conclusion though that what I'm looking for is a mixture of the benefits of CSS-in-JS (one of which is that styles are scoped just like their surrounding JS code, and can be overridden just like their surrounding JS code) and the benefits of Tailwind (which I'm sure you're already familiar with). Hopefully someday we'll get both.
